xr-active-switch.ts

```diff
@@ -0,0 +1,115 @@
+import {Component, Object3D, property} from '@wonderlandengine/api';
+
+export class XrActiveSwitch extends Component {
+    static TypeName = 'xr-active-switch';
+
+    /** Action to perform when XR session starts */
+    @property.enum(['Activate', 'Deactivate', 'Toggle', 'Keep'])
+    ifInXR: string | number = 0;
+
+    /** Action to perform when XR session ends */
+    @property.enum(['Activate', 'Deactivate', 'Toggle', 'Keep'])
+    ifNotInXR: string | number = 1;
+
+    /** Scope of elements to affect */
+    @property.enum(['This Object', 'Children Only', 'This Object & Children'])
+    scope: string | number = 2;
+
+    private components: Component[] = [];
+    private onSessionStart!: () => void;
+    private onSessionEnd!: () => void;
+
+    start() {
+        this.components = [];
+        this.collectComponents();
+
+        // Initial state setup
+        if (this.engine.xr?.session) {
+            this.applyAction(this.ifInXR);
+        } else {
+            this.applyAction(this.ifNotInXR);
+        }
+
+        // Bind event handlers
+        this.onSessionStart = this.handleSessionStart.bind(this);
+        this.onSessionEnd = this.handleSessionEnd.bind(this);
+    }
+
+    onActivate() {
+        this.engine.onXRSessionStart.add(this.onSessionStart);
+        this.engine.onXRSessionEnd.add(this.onSessionEnd);
+    }
+
+    onDeactivate() {
+        this.engine.onXRSessionStart.remove(this.onSessionStart);
+        this.engine.onXRSessionEnd.remove(this.onSessionEnd);
+    }
+
+    private collectComponents() {
+        const mode = this.resolveEnumValue(this.scope, [
+            'This Object',
+            'Children Only',
+            'This Object & Children',
+        ]);
+
+        // Handle current object
+        if (mode === 0 || mode === 2) {
+            const comps = this.object
+                .getComponents()
+                .filter((c) => c.type !== XrActiveSwitch.TypeName);
+            this.components.push(...comps);
+        }
+
+        // Handle children
+        if (mode === 1 || mode === 2) {
+            this.processChildren(this.object);
+        }
+    }
+
+    private processChildren(obj: Object3D) {
+        for (const child of obj.children) {
+            const comps = child
+                .getComponents()
+                .filter((c) => c.type !== XrActiveSwitch.TypeName);
+            this.components.push(...comps);
```

Comment:
```suggestion
            child
                .getComponents()
                .filter((c) => c.type !== XrActiveSwitch.TypeName)
                .forEach(c => this.components.push(c));
```

Avoids allocation (=> garbage collection) of that `comps` array.